### PR TITLE
[CARBONDATA-2876]Fix Avro decimal datatype with precision and scale

### PR DESCRIPTION
**Pull request by Indhumathi27**
**Why is PR for?**

1.Add precision and scale for fieldvalue for Avro Decimal logical type.
2.If Avro schema is of union type with multiple record or multiple enum, then add check for schema.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
     
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by Indhumathi27** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8257/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8260/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/189/

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 1]
```diff
@@ -407,6 +413,19 @@ private Object avroFieldToObjectForUnionType(Schema avroField, Object fieldValue
...
+        // DECIMAL type is defined in Avro as a BYTE type with the logicalType property
+        // set to "decimal" and a specified precision and scale
+        // As binary type is not supported yet,value will be null
+        if (logicalType instanceof LogicalTypes.Decimal) {
+          BigDecimal decimalValue = new BigDecimal(new String(((ByteBuffer) fieldValue).array(),
+              CarbonCommonConstants.DEFAULT_CHARSET_CLASS));
+          out = (decimalValue.round(
+              new MathContext(((LogicalTypes.Decimal) avroField.getLogicalType()).getPrecision())))
+              .setScale(((LogicalTypes.Decimal) avroField.getLogicalType()).getScale(),
+                  RoundingMode.HALF_UP);
```
Comment (manishgupta88) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: replace bigdecimal conversion with below code at all places in the code
`BigDecimal bigDecimal = new BigDecimal(new String(((ByteBuffer) fieldValue).array(), CarbonCommonConstants.DEFAULT_CHARSET_CLASS)
        .setScale(dimension.getColumnSchema().getScale(), RoundingMode.HALF_UP);`

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/227/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8297/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/6/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/119/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8357/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/287/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/121/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/122/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8360/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/290/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/138/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/306/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8376/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/147/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/155/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8394/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/323/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala [hunk 2]
```diff
@@ -884,4 +962,220 @@ class TestNonTransactionalCarbonTableWithAvroDataType extends QueryTest with Bef
...
+    writer.write(avroRec)
+    writer.close()
+    sql(
+      s"""CREATE EXTERNAL TABLE sdkOutputTable STORED BY
+         |'carbondata' LOCATION
+         |'$writerPath' """.stripMargin)
+    checkExistence(sql("select * from sdkOutputTable"), true, "12.80")
+  }
+
+  test("test logical type decimal with data having greater precision") {
```
Comment (KanakaKumar) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala]: Please add a test to verify boundary values like negative and large big decimal values (Big precission needs long to store unscaled value).

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 3]
```diff
@@ -198,14 +199,35 @@ private Object avroFieldToObject(Schema.Field avroField, Object fieldValue) {
...
             break;
           }
           j++;
         }
         out = new StructObject(values);
         break;
+      case BYTES:
+        // DECIMAL type is defined in Avro as a BYTE type with the logicalType property
+        // set to "decimal" and a specified precision and scale
+        // As binary type is not supported yet,value will be null
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: We should throw unsupported data type exception for bytes. Don't store null.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala [hunk 4]
```diff
@@ -884,4 +962,220 @@ class TestNonTransactionalCarbonTableWithAvroDataType extends QueryTest with Bef
...
+    sql("drop table if exists sdkOutputTable")
+    FileFactory.deleteAllCarbonFilesOfDir(FileFactory.getCarbonFile(writerPath))
+    val schema1 =
+      """{
+        |	"namespace": "test.avro",
+        |	"type": "record",
+        |	"name": "Union_data3",
+        |	"fields": [
+        |      {
+        |		  "name": "emp_id",
```
Comment (KanakaKumar) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala]: Please use related names for field. emp_id and union types are not related

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 5]
```diff
@@ -218,9 +240,10 @@ private Object avroFieldToObject(Schema.Field avroField, Object fieldValue) {
    *
    * @param type
    * @param fieldValue
+   * @param unionField
    * @return
    */
-  private boolean checkFieldValueType(Schema.Type type, Object fieldValue) {
+  private boolean checkFieldValueType(Schema.Type type, Object fieldValue, Schema unionField) {
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: Rename method to validateUnionFieldValue()

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala [hunk 6]
```diff
@@ -728,14 +748,46 @@ class TestNonTransactionalCarbonTableWithAvroDataType extends QueryTest with Bef
...
+      .asInstanceOf[ByteBuffer].array(),
+      CarbonCommonConstants.DEFAULT_CHARSET_CLASS)
+    val bytesValue = ByteBuffer.wrap(DatatypeConverter.parseBase64Binary(jsonData))
+    val mySchema =
+      """
+        |{"name": "address",
+        | "type": "record",
+        | "fields": [
+        |  { "name": "street", "type": "string"},
+        |  { "name": "city", "type": {"type" : "bytes",
```
Comment (KanakaKumar) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala]: city and decimal is not correct example

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 7]
```diff
@@ -525,6 +559,11 @@ private static Field prepareFields(Schema.Field avroField) {
           int precision = ((LogicalTypes.Decimal) childSchema.getLogicalType()).getPrecision();
           int scale = ((LogicalTypes.Decimal) childSchema.getLogicalType()).getScale();
           return new Field(fieldName, DataTypes.createDecimalType(precision, scale));
+        } else if (logicalType == null && childSchema.getObjectProp("logicalType")
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: 1) Why this specific check is required?
If this case is for invalid schema, AVRO should validate and throw error. If avro consideres this case just like Bytes, we can also consider same. carbon needs not have this specific check.

2) In the else case we should throw exception or leave to default case to throw bytes is not supported exception.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 8]
```diff
@@ -621,6 +660,11 @@ private static StructField prepareSubFields(String fieldName, Schema childSchema
           int precision = ((LogicalTypes.Decimal) childSchema.getLogicalType()).getPrecision();
           int scale = ((LogicalTypes.Decimal) childSchema.getLogicalType()).getScale();
           return new StructField(fieldName, DataTypes.createDecimalType(precision, scale));
+        } else if (logicalType == null && childSchema.getObjectProp("logicalType")
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: Consider the previous comment in this case also.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/175/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8414/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/343/

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 9]
```diff
@@ -198,14 +199,35 @@ private Object avroFieldToObject(Schema.Field avroField, Object fieldValue) {
...
             break;
           }
           j++;
         }
         out = new StructObject(values);
         break;
+      case BYTES:
+        // DECIMAL type is defined in Avro as a BYTE type with the logicalType property
+        // set to "decimal" and a specified precision and scale
+        // As binary type is not supported yet,value will be null
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: Please remove comment. Its misleading.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java [hunk 10]
```diff
@@ -407,6 +423,24 @@ private Object avroFieldToObjectForUnionType(Schema avroField, Object fieldValue
           out = null;
         }
         break;
+      case BYTES:
+        // DECIMAL type is defined in Avro as a BYTE type with the logicalType property
+        // set to "decimal" and a specified precision and scale
+        // As binary type is not supported yet,value will be null
+        if (logicalType instanceof LogicalTypes.Decimal) {
+          BigDecimal dataValue = new BigDecimal(new BigInteger(((ByteBuffer) fieldValue).array()),
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/AvroCarbonWriter.java]: Please extract logic of constructing the BigDecimal value and validation to a separate method and reuse in both places

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala [hunk 11]
```diff
@@ -884,4 +962,310 @@ class TestNonTransactionalCarbonTableWithAvroDataType extends QueryTest with Bef
...
+    writer.close()
+    sql(
+      s"""CREATE EXTERNAL TABLE sdkOutputTable STORED BY
+         |'carbondata' LOCATION
+         |'$writerPath' """.stripMargin)
+    checkAnswer(sql("select * from sdkOutputTable"),
+      Seq(Row(Row(Row(null,null),Row("ab")))))
+  }
+
+  test("test union with multiple Enum type") {
```
Comment (KanakaKumar) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTableWithAvroDataType.scala]: Can you add test case to read using data source file format. (Syntax "using carbon" with schema Refer SparkCarbonDataSourceTest ).
This will help users on how to define schema for avro logcial types & union .

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/178/

**Comment by KanakaKumar** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/179/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8418/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/347/

**Comment by manishgupta88** [comment]
LGTM
